schema.sql

```diff
@@ -13,258 +18,301 @@ drop table if exists areas;
 
--- only generation units
-create table  generation_units (
-	unit_id integer primary key,
-	name text not null unique,
-	prime_mover text not null,
-	fuel_type text not null,
-	balancing_topology text not null references balancing_topologies(name),
-	start_year integer not null check (start_year >= 0),
-	rating float not null check (rating > 0 ),
-	base_power float not null check (base_power > 0),
-	check (base_power >= rating),
-	foreign key (prime_mover, fuel_type) references prime_mover_types(prime_mover, fuel_type)
+-- Table of entities gets populated automatically once there is an insert on the core tables.
+-- NOTE: This table should not be interacted directly
+create table entities(
+        id integer primary key
+        ,entity_type text not null
+        ,entity_id integer not null
+        ,unique (id, entity_id, entity_type)
 );
 
--- only storge units
-create table  storage_units (
-	storage_unit_id integer primary key,
-	name text not null unique,
-	prime_mover text not null,
-	fuel_type text not null,
-	max_capacity float not null check (max_capacity > 0),
-	round_trip_efficiency float check (round_trip_efficiency >= 0),
-	balancing_topology text not null references balancing_topologies(name),
-	charging_efficiency float not null check (charging_efficiency > 0),
-	discharge_efficiency float not null check (discharge_efficiency > 0),
-	start_year integer not null check (start_year >= 0),
-	rating float not null default 1 check (rating > 0 ),
-	base_power float not null check (base_power > 0),
-	scenario text null,
-	check (base_power >= rating),
-	foreign key (prime_mover, fuel_type) references prime_mover_types(prime_mover, fuel_type)
-);
+-- Table of generation units (aka, existing units).
+create table generation_units (
+   id integer primary key
+  ,name text not null
+  ,prime_mover text not null
+  ,fuel text null
+  ,balancing_topology text not null references balancing_topologies (name)
+  ,start_year integer not null check (start_year >= 0)
+  ,rating real check (rating > 0) default 1.0
+  ,base_power real not null check (base_power > 0)
+  ,check (base_power >= rating)
+  ,foreign key (prime_mover, fuel) references prime_movers (prime_mover, fuel)
+  ,unique (id, name)
+) strict;
 
--- create table for technologies, perhaps change the name to investment_technologies
+-- Table of energy storage units.
+create table storage_units (
+  id integer primary key
+  ,name text not null
+  ,prime_mover text not null
+  ,max_capacity real not null check (max_capacity > 0) -- Energy capacity
+  ,balancing_topology text not null references balancing_topologies (name)
+  ,charging_efficiency real check (
+    charging_efficiency > 0
+    and charging_efficiency <= 1.0
+  ) default 1.0
+  ,discharge_efficiency real check (
+    discharge_efficiency > 0
+    and discharge_efficiency <= 1.0
+  ) default 1.0
+  ,round_trip_efficiency  real generated always as (charging_efficiency * discharge_efficiency) virtual
```

Comment:
In this case, we'll never actually populate `round_trip_efficiency` ourselves, right?

What do we do when we only have 1 efficiency number in Sienna or elsewhere?

Reply:
Correct.  We could make it such that people pass it. I think in some models just take the `round_trip_efficiency` and do not break it down by charge/discharge. 

If we only have one efficiency number we just pass it to w/e it belongs (charging or discharging)

Reply:
why keep round_trip_efficiency ?